design/KEPS/multicluster/proposal.md

```diff
@@ -0,0 +1,160 @@
+
+Some services/workloads may span across clusters but some may not. Business layer services can optionally accept a list of specific clusters to query against for a given s/w/a as an optimization. The business services will need to handle this aggregation properly and cannot assume that s/w/a will exist in all clusters or only in a single cluster.
+
+## UI/UX
+
+This proposal focuses mostly on the backend changes necessary to enable a multi-cluster deployment of Kiali however some UI elements can and should change with a multi-cluster deployment. Exact details of UI changes will be left to future issues.
+
+## Configuration
+
+To deploy a multi-cluster Kiali, users are expected to configure Kiali with access to:
+
+- Centralized prometheus
+- Centralized Jaeger
+- 1..N Kube API servers where N is the number of kube clusters in the mesh. Kiali will require kube api credentials to communicate with each of these kube api servers.
+
+Access to the Kube API servers can be granted by:
+
+1. Creating a service account in the target cluster.
+2. Granting proper permissions for this service account.
+3. Creating a secret with the credentials for this service account in the cluster/namespace where Kiali is deployed.
+
+This process will need to be repeated for every new cluster added. Adding or removing clusters will also require restarting the Kiali pod in order for the cache watches to pick up the new kube config. Adding/removing kube clusters without having to restart Kiali can be an improvement made later.
+
+## Operator
+
+The Kiali operator will still be used for configuring Kiali itself however, it's assumed that the operator won't have access to other kube clusters other than where Kiali is deployed and therefore configuring multicluster access for Kiali will be a manual operation. The operator will remain unchanged in a multi-cluster deployment.
+
+## Testing
+
+All automated e2e testing, API and cypress, is currently performed against a single cluster in CI. There will be a large number of Kiali features related to multi-cluster so there needs to be a way to test multi-cluster deployments frequently and in an automated fashion. Any e2e tests that require a multi-cluster environment will need to be run separately from the rest of the suite.
+
+## Other solutions
+
+### Kiali Agent
+
+A process that runs on each kube cluster that is a part of the mesh and exports Kiali’s view of that mesh to a centralized Kiali. This option offers potential performance gains and ease of configuration since Kiali no longer has to communicate with each individual kube cluster. However, it complicates Kiali’s ability to perform authorization with RBAC since the agents would be exporting data to a centralized Kiali and the centralized Kiali would not have access to the individual clusters. The agent itself adds an additional component to Kiali that would need to be maintained and configured across clusters.
```

Comment:
The more I think about this Kiali agent idea, the more it scares me. It will completely circumvent k8s RBAC security. Yes, the agent itself will have k8s RBAC, but it would expose data to the central Kiali without any k8s RBAC behind it. This will cause angst with many people, and surely would be ripe for CVEs as security holes are punched into the implementation. We should rely on k8s itself for our RBAC/permissions everywhere.

Reply:
Yeah it would either be less secure or add a bunch of complexity that probably isn't even feasible for Kiali to manage. There's probably some users for which having a central Kiali without RBAC enforced on the remote cluster resources is ok but this won't work today and we'd need additional changes to support this less secure model. I'd rather start with a model where RBAC is still enforced on the remote clusters.